lib/cn/ltok.h

```diff
@@ -0,0 +1,102 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+/*
+ * Copyright (C) 2022 Micron Technology, Inc.  All rights reserved.
+ */
+
+#ifndef HSE_CN_LTOK_H
+#define HSE_CN_LTOK_H
+
+#include <hse_util/atomic.h>
+#include <hse_util/inttypes.h>
+#include <hse_util/spinlock.h>
+
+/* MTF_MOCK_DECL(ltok) */
+
+enum ltok_token {
+    ltok_spill,
+    ltok_kcomp,
+    ltok_kvcomp,
+    ltok_split,
+    ltok_join
+};
+
+struct ltok {
+    spinlock_t  ltok_lock;
+    atomic_int *ltok_signal;
+    uint8_t     ltok_active_spills;
+    uint8_t     ltok_active_kcomps;
+    uint8_t     ltok_active_kvcomps;
+    bool        ltok_active_join;
+    bool        ltok_active_split;
+    bool        ltok_reserved;
```

Comment:
A simple boolean may not be enough for reservation. It could cause a deadlock.

Say we kick off spill S1 and then spill S2. But S2 gets to a certain leaf node before S1, enqueues its kvset on the leaf and moves on. The leaf node's spill token won't be dropped until the enqueued S2-kvset is added to it. Now, before S1 gets there, if a split reserves the node, the ltok_reserved flag will prevent S1 from getting the token, while S2-kvset waits for S1.